### ISSUE TITLE
When using @JsonSubTypes and @JsonTypeInfo with multiple inheritance, Jackson is expecting certain order, so ordering of key value pairs in payload matter for Jackson

### ISSUE THREAD

**Issue by Imvikram99** [FasterXML/jackson-databind#3956]
**Describe the bug**
When using @JsonSubTypes and @JsonTypeInfo with multiple inheritance, Jackson is not able to convert JSON into Java objects

**Version information**
Which Jackson version(s) was this for?
com.fasterxml.jackson.core:jackson-databind:2.13.0

**To Reproduce**
If you have a way to reproduce this with:

1. create class : 
2. @JsonTypeInfo(
        use = JsonTypeInfo.Id.NAME,
        include = JsonTypeInfo.As.PROPERTY,
        property = "type")
@JsonSubTypes({
        @JsonSubTypes.Type(name = "pokerMove", value = PokerMove.class),
        // Add other subclasses if any
})
public class Move {
    private String playerId; // ID of the player making the move
    private MoveTypeInterface moveTypeInterface;
}

2.2) another class:

@JsonTypeInfo(
        use = JsonTypeInfo.Id.NAME,
        include = JsonTypeInfo.As.PROPERTY,
        property = "type")
@JsonSubTypes({
        @JsonSubTypes.Type(name = "pokerMoveType", value = PokerMoveType.class),
        @JsonSubTypes.Type(name = "moveType", value = MoveType.class),
        @JsonSubTypes.Type(name = "chessMove", value = ChessMoveType.class),
        @JsonSubTypes.Type(name = "cricketMove", value = CricketMoveType.class)
})
public interface MoveTypeInterface {
    Enum<?> getMoveType();
    int getValue();
}

2.3)public enum PokerMoveType implements MoveTypeInterface {
    BET("BET"), RAISE("RAISE"), CALL("CALL"), FOLD("FOLD"), CHECK("CHECK"), ALL_IN("ALL_IN"),YET_TO_MAKE_MOVE("YET_TO_MAKE_MOVE");
    @JsonValue
    private String name;

    @JsonCreator
    public static PokerMoveType fromJson(String name) {
        for (PokerMoveType pokerMoveType : PokerMoveType.values()) {
            if (pokerMoveType.name.equals(name)) {
                return pokerMoveType;
            }
        }
        return null;
    }

    private PokerMoveType(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    @Override
    public Enum<?> getMoveType() {
        return this;
    }

    @Override
    public int getValue() {
        return 0;
    }
}
2.4)
public class PokerMove extends Move {
    private int betAmount; // The amount the player is betting
    private int potSize; // The current size of the pot

    public PokerMove(String playerId, MoveTypeInterface moveType, int betAmount, int potSize) {
        super(playerId, moveType);
        this.betAmount = betAmount;
        this.potSize = potSize;
        log.info("Poker move created: Player ID: {}, Move Type: {}, Bet Amount: {}, Pot Size: {}",
                playerId, moveType, betAmount, potSize);
    }

}

2.5) try to get move in your controller
   public Game makeMove(@PathVariable String gameId, @RequestBody Move move) {}
   
2.6) hit the api using payload: {
    "playerId": "player4",
    "type": "pokerMove",
    "betAmount": 10,
    "potSize": 1,
     "moveTypeInterface": {
        "type": "pokerMoveType",
        "BET": "10"
    }
}

2.7) everything works, you get the every data but when you try with below payload:
{
    "playerId": "player4",
    "type": "pokerMove",
     "moveTypeInterface": {
        "type": "pokerMoveType",
        "BET": "10"
    },
    "betAmount": 10,
    "potSize": 1
   
}
you won't get potsize value in your request , it will default to 0.

3. Longer example stored somewhere else (diff repo, snippet), add a link
4. Textual explanation: include here

**Expected behavior**
order of key value pairs inside payload shouldn't matter

**Additional context**

**Comment by pjfanning** [comment]
Could you try a newer Jackson release eg 2.15.1? We sort of need a 2.15.2 to sort out some issues but it would be useful to know how latest code performs.

**Comment by Imvikram99** [comment]
i m facing some issues , towards failOnRepeatedNames = true in the newer versions. so unable to test

**Comment by cowtowncoder** [comment]
Version 2.13 no longer supported so reproduction on 2.15 is required. But I assume problem still exists.

Also note that reproduction needs to be stand-alone (cannot rely on Spring or other frameworks). I assume this is not a problem here and reproduction without such dependencies would work.

**Comment by cowtowncoder** [comment]
On version: use of `2.13.0` makes no sense either: it is strongly recommended to always use the latest patch version of minor version.
So at least try `2.13.5` which has same defaults. There are many fixes between 2.13.0 and 2.13.5 so if there was a problem with content buffering (wrt field order having effect), that may have been resolved.

**Comment by Imvikram99** [comment]
@cowtowncoder thanx for your valuable time and comment, yes i could try with 2.13.5 but same issue exists there as well.

**Comment by Imvikram99** [comment]
@cowtowncoder : please find the test case here:

package ai.apiverse.mars;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.fasterxml.jackson.annotation.JsonValue;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class Test {

    public static void main(String[] args) throws Exception {
        ObjectMapper mapper = new ObjectMapper();
        mapper.enable(SerializationFeature.INDENT_OUTPUT);
        // from payload to java class i need to make, need to provide base class properties along with a field type with value derived1

       // String jsonPayloadWithValueD1CorrectlyDeserialsed = "{\"id\": 1, \"name\": \"Test\", \"type\": \"derived1\", \"d1\": 10, \"moveTypeInterface\": {\"type\": \"pokerMoveType\", \"value\": 10}}";
        String jsonPayload = "{\"id\": 1, \"name\": \"Test\", \"type\": \"derived1\", \"moveTypeInterface\": {\"type\": \"pokerMoveType\", \"value\": 10}, \"d1\": 10}";

        Base base = mapper.readValue(jsonPayload, Base.class);

        assertEquals(1, base.getId());
        assertEquals("Test", base.getName());

        // Check the deserialization for Derived1
        assertEquals(Derived1.class, base.getClass());
        Derived1 derived1 = (Derived1) base;
        assertEquals(10, derived1.getD1());

        // Check the deserialization for MoveTypeInterface
        assertEquals(PokerMoveType.class, base.getMoveTypeInterface().getClass());
        PokerMoveType moveType = (PokerMoveType) base.getMoveTypeInterface().getMoveType();
        assertEquals(10, base.getMoveTypeInterface().getValue());
    }

}

@JsonTypeInfo(
        use = JsonTypeInfo.Id.NAME,
        include = JsonTypeInfo.As.PROPERTY,
        property = "type")
@JsonSubTypes({
        @JsonSubTypes.Type(name = "derived1", value = Derived1.class),
        // Add other subclasses if any
})
class Base {

   // @JsonTypeInfo(use = Id.CLASS, include = JsonTypeInfo.As.PROPERTY, property = "type")
    private int id;

    private String name;
    private MoveTypeInterface moveTypeInterface;

    public void setMoveTypeInterface(MoveTypeInterface moveTypeInterface){
        this.moveTypeInterface = moveTypeInterface;
    }

    public MoveTypeInterface getMoveTypeInterface(){
        return this.moveTypeInterface ;
    }
    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

//    public Derived1 getDerived1() {
//        return derived1;
//    }
//
//    public void setDerived1(Derived1 derived1) {
//        this.derived1 = derived1;
//    }
//
//    public Derived2 getDerived2() {
//        return derived2;
//    }
//
//    public void setDerived2(Derived2 derived2) {
//        this.derived2 = derived2;
//    }

}

class Derived1 extends Base {
    int d1;
    public  void setD1(int a){
        this.d1 = a;
    }
    public int getD1(){
        return this.d1;
    }
}

class Derived2 extends Base {

}

@JsonTypeInfo(
        use = JsonTypeInfo.Id.NAME,
        include = JsonTypeInfo.As.PROPERTY,
        property = "type")
@JsonSubTypes({
        @JsonSubTypes.Type(name = "pokerMoveType", value = PokerMoveType.class),
})
 interface MoveTypeInterface {
    Enum<?> getMoveType();
    int getValue();
}
enum PokerMoveType implements MoveTypeInterface {
    BET("BET"), RAISE("RAISE"), CALL("CALL"), FOLD("FOLD"), CHECK("CHECK"), ALL_IN("ALL_IN"), YET_TO_MAKE_MOVE("YET_TO_MAKE_MOVE");
    @JsonValue
    private String name;

    private PokerMoveType(String name) {
        this.name = name;
    }

    @JsonCreator
    public static PokerMoveType fromJson(String name) {
        for (PokerMoveType pokerMoveType : PokerMoveType.values()) {
            if (pokerMoveType.getName().equals(name)) {
                return pokerMoveType;
            }
        }
        return null;
    }

    public String getName() {
        return name;
    }

    @Override
    public Enum<?> getMoveType() {
        return this;
    }

    @Override
    public int getValue() {
        return this.getValue();
    }
}

**Comment by cowtowncoder** [comment]
Thank you @Imvikram99!

**Comment by cowtowncoder** [comment]
Ok there are some issues with the test... for example: this leads to StackOverflowError:

```
        @Override
        public int getValue() {
            return this.getValue();
        }
```

and for serialization this could as well:

```
    @Override
    public Enum<?> getMoveType() {
        return this;
    }
```

But more importantly, JSON suggest for deserialization is likely wrong: when serializing `Enum` with `@JsonValue` annotation serializing as String, result with type info will be an Array value, not Object.
The proper way to test things here would then be serializing full value, deserializing -- Jackson only guarantees it should be able to deserialize polymorphic values it writes and not arbitrary JSON. Ordering of properties should not, however, matter.

So basically test case as is does not really point out actual problem there may be.

**Comment by cowtowncoder** [comment]
Ok, so, `fromJson()` returning `null` is masking the problem: type id passed is empty String. This is probably because given serialization of `MoveTypeInterface` value is wrong; serializer would not produce JSON Object but JSON Array.

**Comment by cowtowncoder** [comment]
I will change handling in `FactoryBasedEnumDeserializer` so that it will not try to get String out of JSON Object but throw Exception -- that would have made it clear here what the problem is, and prevent really odd failure mode: what is happening here is that code is trying to get `String` but noticing there's no scalar value -- but not skipping rest of Object value. And quietly passing id of "" (empty String), which factory method returns as `null`.

But as to your specific problem: JSON being passed as alleged serialization is not compatible with polymorphic type handling. If you can modify test to show that such JSON is being serialized by Jackson I can see what might be causing it?

**Comment by Imvikram99** [comment]
thanks for quick response @cowtowncoder , yes it's expecting array not object , but didn't get why array , while in my class definition i have given it as a java object . also can you please correct the payload for me. it will be really very helpful, for me to understand the problem

**Comment by cowtowncoder** [comment]
@Imvikram99 Serialization will only use Object value if it is technically possible -- but when `Enum` is serialized as String, it cannot -- only if serialization was an Object (like it usually is for POJOs). For other values it must then do something else and that something else is to use `WRAPPER_ARRAY` style instead (where 2-element Array has type id first, and value second; any value shape is fine). In fact, use of Object property cannot work for any scalar values, or Lists, arrays etc, which is why for general purpose serialization use of `WRAPPER_ARRAY` is a better default. But users seem to wan to try to use `PROPERTY` (Object property) more for some reason, regardless of its downsides.

This all works automatically when serializing so what Jackson supports (and only this) is that whatever serialization it (would) use is expected for deserialization. This is also why when troubleshooting it is best to start by serializing what you have or expect and follow that path.

**Comment by Imvikram99** [comment]
thanks @cowtowncoder for your time, you are a genius!! it helped me a lot.

**Comment by cowtowncoder** [comment]
@Imvikram99 Great! Good luck.
